File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {
    PriorityQueue<Integer> leftPartition;
    PriorityQueue<Integer> rightPartition;
    
    public MedianFinder() {
        leftPartition = new PriorityQueue<>((a, b) -> b - a);
        rightPartition = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        if (leftPartition.size() == 0 || leftPartition.peek() > num) {
            leftPartition.add(num);
        } else {
            rightPartition.add(num);
        }
        
        if (Math.abs(leftPartition.size() - rightPartition.size()) > 1) {
            rightPartition.add(leftPartition.poll());
        } else if (leftPartition.size() < rightPartition.size()) {
            leftPartition.add(rightPartition.poll());
        }
    }
    
    public double findMedian() {
        if (leftPartition.size() == rightPartition.size()) {
            return (double)(leftPartition.peek() + rightPartition.peek()) / 2.0;
        }
        return (double)(leftPartition.peek());
    }
}

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int> lf;
    priority_queue<int, vector<int>, greater<int>> rt;
    
    MedianFinder() {}
    
    void addNum(int num) {
        if (lf.empty() || num <= lf.top()) {
            lf.push(num);
        } else {
            rt.push(num);
        }
        
        if (lf.size() > 1 + rt.size()) {
            rt.push(lf.top()); lf.pop();
        } else if (rt.size() > lf.size()) {
            lf.push(rt.top()); rt.pop();
        }
    }
    
    double findMedian() {
        if ((lf.size() + rt.size()) & 1) {
            return double(lf.top());
        } else {
            return double(lf.top() + rt.top()) / 2.0;
        }
    }
};